src/ripple/protocol/impl/tokens.cpp

```diff
@@ -162,90 +180,307 @@ encodeToken (TokenType type,
+            "34444488709877454747081479018851205589883534409595458615323760395288576"};  // 58^40
+        static checked_uint512_t const c58_50{
+            "14838830640714294999983396511843381888381117770762056659674968992599706869653743415066624"};  // 58^50
+        using mpUint128 = DecodeBase58Detail::mpUint<128>;
+        using mpUintMaxOut = DecodeBase58Detail::mpUint<maxOutBits>;
+        bool const secureErase = flags.test(DecodeBase58Detail::maybeSecret);
+        // low result accounts for the first two base 58^10 coefficients. This is at most 2^64+2^62*58^10, or 123 bits.
+        mpUint128 low_result{secureErase};
+        low_result.emplace(b5810[0] + c58_10 * b5810[1]);
+        // high result accounts all but the first two base 58^10 coefficients.
+        mpUintMaxOut high_result{secureErase};
+        switch (b5810i)
+        {
+            case 3:
+                high_result.emplace(b5810[2] * c58_20);
+                break;
+            case 4:
+                high_result.emplace(b5810[2] * c58_20 + b5810[3] * c58_30);
+                break;
+            case 5:
+                high_result.emplace(b5810[2] * c58_20 + b5810[3] * c58_30 + b5810[4] * c58_40);
+                break;
+            case 6:
+                high_result.emplace(b5810[2] * c58_20 + b5810[3] * c58_30 +
+                                    b5810[4] * c58_40 + b5810[5] * c58_50);
+                break;
+            default:
+                high_result.emplace(0);
+                break;
+        }
+        mpUintMaxOut result{secureErase};
+        result.emplace(*low_result + *high_result);
+        DecodeBase58Detail::TempBuf<std::uint8_t, maxOutBytes> tmp{secureErase};
+        memset(tmp.data(), 0, zeroes);
+        // Account for leading zeroes by starting the write at an offset
+        auto endWritten =
+            boost::multiprecision::export_bits(*result, tmp.data() + zeroes, 8);
+        if (zeroes && endWritten == tmp.data() + zeroes + 1 && tmp[zeroes] == 0)
+        {
+            // The last zero is redundant and should be removed
+            --endWritten;
+        }
+        std::size_t const numWritten = std::distance(tmp.data(), endWritten);
+        assert(numWritten <= maxOutBytes);
+        tmp.numToErase = numWritten;
+        if (numWritten <= 4)
+            return {};
+        if (numWritten > MaxDecodedTokenBytes)
+            return {};
+
+        if (out.size())
+            memset(out.data(), 0, out.size());
+        DecodeMetadata metadata;
+        memset(metadata.encodingType.data(), 0, metadata.encodingType.size());
+        memcpy(metadata.checksum.data(), tmp.data() + numWritten - 4, 4);
+        metadata.tokenType = tmp[0];
+        std::uint8_t const* dataStart = &tmp[1];
+        std::uint8_t const* dataEnd = &tmp[0] + numWritten - 4;
+        if (flags.test(DecodeBase58Detail::maybeRippleLibEncoded) &&
+            static_cast<TokenType>(metadata.tokenType) == TokenType::None &&
```

Comment:
I see those same constants again... 0xE1 and 0x4B... along with lots of other magic numbers. `dataEnd - dataStart == 18`?

Reply:
Addressed in 91c6c712b0973af30980b849d50e33a4f715d449